### DISCUSSION TITLE
How to delete a specific element from an observable collection

### DISCUSSION THREAD

**Discussion by reelthymeos** [AvaloniaUI/Avalonia#17302] [Q&A]
Hi,

I have list of items that are displayed in a View and there are options to add an item to the list, clear all items, and delete a specific item. 

I have not been able to successfuly implement the functionality of deleting specific items. I have referenced the Avalonia Sample "ToDo list", which shows how to do this, though I should note that in my project, the ViewModel is a ReactiveObject and not based off the MVVM community toolkit. I started using ReactiveObjects because I wanted a way to bind buttons to commands in the ViewModel and that was the example that was shown to me. 

Here is the underlying 'model', called step:

```
public enum Units
	{
		ms = 1,
		s = 1000,
		min = 60000,
		hour = 3600000
	}

	public class Step : INotifyPropertyChanged
	{
	
		public string SelectedUnits
		{
			get
			{
				return nameof(MyUnits);
			}

			set
			{
				MyUnits = (Units)Enum.Parse(typeof(Units), value);
				RaisePropertyChanged(nameof(SelectedUnits));
			}
		}

		public Units MyUnits { get; private set; }
		public List<string> UnitList { get; } = new List<string>() { "ms", "s", "min", "hour" };
		

		public Step()
		{
		}

		// This event is implemented by "INotifyPropertyChanged" and is all we need to inform
		// our view about changes.
		public event PropertyChangedEventHandler? PropertyChanged;

		private void RaisePropertyChanged([CallerMemberName] string? propertyName = null)
		{
			PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
		}
	}

```

Here is the ViewModel:
```


public class StepExecutorViewModel : ReactiveObject
	{
		private ObservableCollection<Step> steps = new ObservableCollection<Step>();

		public ICommand AddStep { get; }
		public ICommand DeleteStep { get; }
		public ICommand ClearSteps { get; }

		public StepExecutorViewModel()
		{
			AddStep = ReactiveCommand.Create(OnAddStep);
			DeleteStep = ReactiveCommand.Create(OnDeleteStep);
			ClearSteps = ReactiveCommand.Create(OnClearSteps);
		}
		
		public ObservableCollection<Step> Steps
		{
			get { return steps; }
		}

		public void OnAddStep()
		{
			//action to add a step after pressing 'add step' button
			steps.Add(new Step());
		}

		public void OnClearSteps()
		{
			//clear all steps from the list
			steps.Clear();
		}

		public void OnDeleteStep()
		{
			//I am supposed to feed the specific instance of 'step' here as a parameter, then delete from list?
			// problem is that reactive commands don't like that	
		}

	}
```

The problem in this above approach is that ReactiveCommand.Create() doesn't like when the method that I provide it is not a System.Action, which apparently it can't be if there is a parameter in the function? It throws the error "Argument 1: cannot convert from 'method group' to 'System.Action'".

The second issue/question I have is in regards to the View. I know that in order to specify the step I want to delete, I need to pass the specific step as a Command Parameter. I am just having a hard time understanding all the syntax related to casting the DataContext that needs to happen. So in this case, how can I properly populate the command and command parameter fields? i.e. how do I express the bindings for these?  Below is my View:

```
<DataGrid Grid.Row="6" Height="200" Grid.ColumnSpan="5" 
					  FontWeight="Bold"
					  VerticalScrollBarVisibility="Visible"
					  ItemsSource="{Binding Steps}" >

					<DataGridTemplateColumn Width="Auto" Header="Units">
						<DataGridTemplateColumn.CellTemplate>
							<DataTemplate DataType="cons:Step">
								<StackPanel Orientation="Horizontal">
								<ComboBox Margin="5" Width="100" ItemsSource="{Binding UnitList}"/>
								<Button ToolTip.Tip="Delete Step" Margin="5" Command="{Binding DeleteStep}" Content="X" />
								</StackPanel>
							</DataTemplate>
						</DataGridTemplateColumn.CellTemplate>
					</DataGridTemplateColumn>

				</DataGrid.Columns>

</DataGrid>
```

I tried to do the binding for the delete button like this (taken from Avalonia ToDo list example:

```
<Button Command="{Binding #Root.((vm:StepExecutorViewModel)DataContext).DeleteStep}
CommandParameter="{Binding .}"
</Button>
```

When I run this, I get the following exception: 

>  "An object implementing IHandleObservableErrors (often a ReactiveCommand or ObservableAsPropertyHelper) has errored, thereby breaking its observable pipeline. To prevent this, ensure the pipeline does not error, or Subscribe to the ThrownExceptions property of the object in question to handle the erroneous case.",

Does anyone have some insight into what I could be doing wrong here?

**Comment by thevortexcloud** [comment]
Just do `CommandParameter="{Binding }"` on your button and it will probably work. Make sure your command handler also accepts an argument.

**Reply by reelthymeos** [reply]
In my post above I mentioned that I get a compile error when trying to have a parameter in my command handler. Here is the issue:
 

> The problem in this above approach is that ReactiveCommand.Create() doesn't like when the method that I provide it is not a System.Action, which apparently it can't be if there is a parameter in the function? It throws the error "Argument 1: cannot convert from 'method group' to 'System.Action'".

**Reply by thevortexcloud** [reply]
ReactiveUI often requires you to specify the input and output types. This is one of the examples in the official docs:

> ReactiveCommand<int,Unit> command = ReactiveCommand.Create<int>(
    integer => Console.WriteLine(integer));

A `Unit` is the Reactive UI equivalent of void.

https://www.reactiveui.net/docs/handbook/commands/

**Comment by timunie** [comment]
In Avalonia.Samples there is a new ToDoList sample which does similar to this. Check it out.

**Reply by timunie** [reply]
Ps OnDeleteStep is missing a paramter. 🤗

**Reply by reelthymeos** [reply]
In my post above I mentioned that I get a compile error when trying to have a parameter in my command handler. Here is the issue:

The problem in this above approach is that ReactiveCommand.Create() doesn't like when the method that I provide it is not a System.Action, which apparently it can't be if there is a parameter in the function? It throws the error "Argument 1: cannot convert from 'method group' to 'System.Action'".

**Reply by timunie** [reply]
https://github.com/AvaloniaUI/Avalonia.Samples/blob/main/src/Avalonia.Samples/MVVM/CommandSample/ViewModels/MainWindowViewModel.cs

**Reply by reelthymeos** [reply]
I now fixed it to include the parameter in the command handler and I redefined the command as such:

```
public ICommand DeleteStep { get; }
DeleteStep = ReactiveCommand.Create<Step>(step => OnDeleteStep(step));
```

This gets rid of the compile error. However, since this was added, If I now hit any of the buttons, including the one to add a step, which triggers the AddStep command, I get this error:

> System.ArgumentException: 'Unable to resolve type vm:StepExecutorViewModel from any of the following locations: '

I'm thinking it may not like this line, because this error did not happen before I included it:

`<Button Command="{Binding #Root.((vm:StepExecutorViewModel)DataContext).DeleteStep}" CommandParameter="{Binding }" Content="X" />
`
In my view, I have the following defined:

`xmlns:vm="using:MyApp.ViewModels"`

and the view model StepExecutorViewModel is this view's datacontext. Various other buttons and controls are able to effectively bind to properties in StepExecutorViewModel.

I know that since this button is defined inside a DataTemplate, that is no longer the DataContext, but rather the dataContext is the DataType (the model, not ViewModel). So I believe that line is to make it so that the DataContext is the parent ViewModel and not the Model. Could something about the syntaxt be wrong in the casting?
